shachain/element.go

```diff
@@ -0,0 +1,134 @@
+package shachain
+
+import (
+	"crypto/sha256"
+	"errors"
+
+	"github.com/roasbeef/btcd/wire"
+)
+
+// Element represent the entityt which contains the hash and
+// correponding to it index. By comparing two indexes we may change hash in
+// such way to derive another element.
+type Element struct {
+	index index
+	hash  *wire.ShaHash
+}
+
+// NewElementFromStr creates new element by the given hash string.
+func NewElementFromStr(s string, index index) (*Element, error) {
+	hash, err := hashFromString(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Element{
+		index: index,
+		hash:  hash,
+	}, nil
+}
+
+// derive is used to get the one shachain element from another by applying
+// bits flipping and hash operation based on index.
+func (e *Element) derive(to index) (*Element, error) {
+	positions, err := e.index.derive(to)
+	if err != nil {
+		return nil, err
+	}
+
+	buf := e.hash.Bytes()
+	for _, position := range positions {
+		// Flip the bit and then hash the current state.
+		h := sha256.Sum256(changeBit(buf, position))
```

Comment:
I'm still working through this file and the entirety of this pull request, but I don't think something as simple as a bit flipping operation should be hidden behind a function. Instead, inlining the logic makes the code easier to follow as the reader doesn't nee d to jump around the file to confirm her expectation of the function based on its name.